### DISCUSSION TITLE
Could OpenJpegProcessor support streaming sources?

### DISCUSSION THREAD

**Discussion by donaldgray** [cantaloupe-project/cantaloupe#563] [Q&A]
The documentation for [`OpenJpegProcessor`](https://cantaloupe-project.github.io/manual/5.0/processors.html#OpenJpegProcessor) states that: 

> This processor can only read from files. For use with sources other than FilesystemSource, one of the fallback retrieval strategies must be used.

Is the issue that the OpenJpeg library doesn't support streaming; or is the `OpenJpegProcessor` itself that is unable to stream? 

For the latter I can see that there is an open PR (https://github.com/uclouvain/openjpeg/pull/1212) in OpenJpeg which seems to add support for buffer-based streams. However, after reading https://github.com/cantaloupe-project/cantaloupe/issues/561#issuecomment-1055925759 it seems like there needs to be work done in _both_ the OpenJpeg lib and `OpenJpegProcessor`? Or is there a fundamental issue that would prevent `OpenJpegProcessor` from ever supporting streaming?

**Comment by adolski** [comment]
OpenJpegProcessor invokes OpenJPEG's opj_decompress tool to decode JP2s, and opj_decompress can only read from files. This is one of several important limitations of OJP. OJP needs to be scrapped and replaced with an Image I/O plugin that calls into libopenjp2, which already supports non-file sources.

**Reply by donaldgray** [reply]
Thanks for the feedback. Do you know if there are plans to rewrite of OJP at all?

If there are no plans to rewrite is it something that _could_ be done by someone that's not familiar with the codebase?

**Reply by adolski** [reply]
There are plans, but they are pretty far off still. I'm waiting on the features in [Project Panama](https://openjdk.java.net/projects/panama/) to make it into a JRE release, and also on finding some free time.

This would be done in the context of #388, so it wouldn't require any knowledge of Cantaloupe's code base.

I'm aware of a similar Image I/O plugin that calls natively into OpenJPEG, using JNR instead of Panama, but I haven't used it: https://github.com/dbmdz/imageio-jnr

That plugin might help you right away. The latest Cantaloupe can theoretically use non-bundled Image I/O plugins like it with Java2dProcessor via the `processor.imageio.*` config keys, but this feature is not fully fleshed out or tested, and I don't know if it will work.
